src/preact/hooks/useMutationObserver.tsx

```diff
@@ -0,0 +1,51 @@
+import {useEffect, useMemo} from '#preact';
+import useEvent from '#preact/hooks/useEvent';
+import {RefObject} from '#preact/types';
+
+/**
+ * Subscribes to an element's DOM mutations
+ */
+export function useMutationObserver(
+  elementRef: RefObject<HTMLElement>,
+  config: MutationObserverInit,
+  onMutation: (record: MutationRecord) => void
+) {
+  const onMutationHandler = useEvent(onMutation);
+  const mo = useMemo(() => {
+    return new MutationObserver((records) => {
+      records.forEach((record) => {
+        onMutationHandler(record);
+      });
+    });
+  }, [onMutationHandler]);
+
+  useEffect(() => {
+    if (elementRef.current) {
+      mo.observe(elementRef.current, config);
+      return () => mo.disconnect();
+    }
+  }, [config, /** These are stable: */ elementRef, mo]);
```

Comment:
This is a little dangerous because the `config` object could be a new instance of an object and this doesn't do deep checking.

Reply:
Agreed ... in the `useAttributeObserver`, I have to use `useMemo` to ensure it stays "stable".  
We don't currently have a "deep-check" utility, and since this config has a lot of potential fields, it's not super easy to memoize.  
The expected use-case for this option is a top-level constant, like:
```
const observeExpandedAttribute = { attributes: true, { attributeFilter: [ 'expanded' ] } };
function Component() {
  useMutationObserver(ref, observeExpandedAttribute, () => { ... });
}
```

So, I figure we have these options:
1. Leave it (exposing the risk that an "inline" configuration will cause `useEffect` thrash)
2. Omit `config` from the dependencies array, and never re-connect if it changes.  (I'd have to use `useRef` to get eslint to pass) 
3. Create a deep-compare version of `useMemo`. Not super hard, but ugly.

Which do you like best?

Reply:
I went with option 3. I added a `useDeepValue(value)` hook which does a deep-compare and returns the previous value when there's a match.  

If the object reference is constant (like in my 2 examples) then there's no performance impact.
If the object reference is always changing, then there's a minimal deep-check but no `useEffect` thrashing.  
So it's kinda the best of both worlds!